README.md
# contentserver

A Clojure library designed to mimic a bank account deposit, withdrawal and balance actions.

## Usage

Download the project into your local computer. Alternatively you can clone the project by opening a Linux terminal (or Git BASH emulator for Windows) and run the below command:

git clone https://github.com/abalajoe/bankaccount

For UNIX based users, open the terminal, navigate to the run folder inside the project folder. Next, make the start script executable by executing the below command:

    sudo chmod +x startup.sh

Now you can run the program by executing the below command:

    sudo ./startup.sh

Alternatively (without using startup.sh), open the terminal, navigate to the run folder and execute the below commands:

    java -Dlog4j.configuration=file:"log4j.properties" -Dbankaccount.config="config.properties" -jar bankaccount.jar

For windows based users, navigate to the run folder, open the startup.bat file and edit the Java path to suit your computer's Java path. Next, open the command prompt, navigate to the run folder and execute the below command:

    startup.bat

NB: You must have Java installed on your computer and internet connection (Datasource is on cloud) to run the application.


## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-02-04
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-02-04
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/contentserver/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/contentserver/compare/0.1.0...0.1.1


test/contentserver/core_test.clj
(ns contentserver.core-test
  ^{:author "jabala"
    :doc "Bank Account API tests"
    :added "1.0"
    }
  (:use midje.sweet)
  (:require [clojure.test :refer :all]
            [contentserver.core :refer :all]
            [clj-http.client :as http]
            [ring.mock.request :as mock]
            [contentserver.httpserver :as httpserver]
            [clojure.data.json :as json]
            [contentserver.model.datasource :as db]))

(facts "General Account testing"
       (fact "Default route"
             (let [response (httpserver/app (mock/request
                                              :get "/"
                                              ))] ;
               (:status response) => 200
               (:headers response) => {"Content-Type" "application/json; charset=utf-8", "X-Content-Type-Options" "nosniff", "X-Frame-Options" "SAMEORIGIN", "X-XSS-Protection" "1; mode=block"}
               (:body response) => (json/write-str {:status-code 0 :status-msg "Welcome to Bank Account API"})))
       (fact "Check Balance"
             (let [response (httpserver/app (mock/request
                                              :get "/balance"
                                              ))] ;
               (:status response) => 200
               (:headers response) => {"Content-Type" "application/json; charset=utf-8", "X-Content-Type-Options" "nosniff", "X-Frame-Options" "SAMEORIGIN", "X-XSS-Protection" "1; mode=block"}
               (:body response) => (json/write-str {:status-code 0 :status-msg "Your balance request is successful"
                                                    :balance (:balance (first (db/get-account-information)))})))
       (fact "Invalid Route"
             (let [response (httpserver/app (mock/request
                                              :get "/invalid"
                                              ))] ;
               (:status response) => 404
               (:headers response) => {"Content-Type" "application/json; charset=utf-8", "X-Content-Type-Options" "nosniff", "X-Frame-Options" "SAMEORIGIN", "X-XSS-Protection" "1; mode=block"}
               (:body response) => (json/write-str {:status-code -1 :status-msg "Invalid Request"}))))

(facts "Deposit Account testing"
       (fact "Invalid input"
             (let [response (httpserver/app (mock/request
                                              :post "/deposit"
                                              {:deposit "hello"}))] ;
               (:status response) => 422
               (:headers response) => {"Content-Type" "application/json; charset=utf-8", "X-Content-Type-Options" "nosniff", "X-Frame-Options" "SAMEORIGIN", "X-XSS-Protection" "1; mode=block"}
               (:body response) => (json/write-str {:status-code -2 :status-msg "Invalid Amount"})))
       (fact "Invalid Amount"
             (let [response (httpserver/app (mock/request
                                              :post "/deposit"
                                              {:deposit "9"}))] ;
               (:status response) => 422
               (:headers response) => {"Content-Type" "application/json; charset=utf-8", "X-Content-Type-Options" "nosniff", "X-Frame-Options" "SAMEORIGIN", "X-XSS-Protection" "1; mode=block"}
               (:body response) => (json/write-str {:status-code -2 :status-msg "Invalid Amount"})))
       (fact "Max Daily Deposit"
             (let [response (httpserver/app (mock/request
                                              :post "/deposit"
                                              {:deposit "50000"}))] ;
               (:status response) => 422
               (:headers response) => {"Content-Type" "application/json; charset=utf-8", "X-Content-Type-Options" "nosniff", "X-Frame-Options" "SAMEORIGIN", "X-XSS-Protection" "1; mode=block"}
               (:body response) => (json/write-str {:status-code -2 :status-msg "You cannot deposit more than $40,000"}))))

(facts "Withdraw Account testing"
       (fact "Invalid input"
             (let [response (httpserver/app (mock/request
                                              :post "/withdraw"
                                              {:deposit "hello"}))] ;
               (:status response) => 422
               (:headers response) => {"Content-Type" "application/json; charset=utf-8", "X-Content-Type-Options" "nosniff", "X-Frame-Options" "SAMEORIGIN", "X-XSS-Protection" "1; mode=block"}
               (:body response) => (json/write-str {:status-code -2 :status-msg "Invalid Amount"})))
       (fact "Invalid Amount"
             (let [response (httpserver/app (mock/request
                                              :post "/withdraw"
                                              {:deposit "9"}))] ;
               (:status response) => 422
               (:headers response) => {"Content-Type" "application/json; charset=utf-8", "X-Content-Type-Options" "nosniff", "X-Frame-Options" "SAMEORIGIN", "X-XSS-Protection" "1; mode=block"}
               (:body response) => (json/write-str {:status-code -2 :status-msg "Invalid Amount"})))
       (fact "Max Daily Withdrawal"
             (let [response (httpserver/app (mock/request
                                              :post "/withdraw"
                                              {:withdraw "50000"}))] ;
               (:status response) => 422
               (:headers response) => {"Content-Type" "application/json; charset=utf-8", "X-Content-Type-Options" "nosniff", "X-Frame-Options" "SAMEORIGIN", "X-XSS-Protection" "1; mode=block"}
               (:body response) => (json/write-str {:status-code -2 :status-msg "You cannot withdraw more than $20,000"}))))

project.clj
(defproject contentserver "0.1.0-SNAPSHOT"

  :description "FIXME: write description"
  :url "http://example.com/FIXME"

  :dependencies [[org.clojure/clojure "1.7.0"]
                 [http-kit "2.1.18"]
                 [org.clojure/tools.logging "0.2.6"] ;; loggging
                 [propertea "1.2.3"]              ;; reading from .properties
                 [org.clojure/java.jdbc "0.3.6"]  ;; jdbc
                 [org.postgresql/postgresql "9.3-1101-jdbc41"] ; postgres driver
                 [korma "0.4.0"]                            ;; db library
                 [org.clojure/data.json "0.2.5"]
                 [log4j "1.2.17"]
                 [compojure "1.1.8"]                        ; routing
                 [ring/ring-codec "1.0.0"]
                 [ring/ring-mock "0.3.0"]
                 [ring "1.3.0"]
                 [cheshire "5.4.0"]
                 [clj-time "0.6.0"]
                 [midje "1.6.3"]
                 [org.slf4j/slf4j-log4j12 "1.6.4"]
                 [ring/ring-defaults "0.1.2"]
                 [com.novemberain/langohr "3.0.1"]
                 [selmer "0.8.0"]
                 [pg-hstore "3.0.3"]
                 [environ "1.0.0"]
                 [markdown-clj "0.9.64"]
                 [lib-noir "0.9.4"]]

  :main ^:skip-aot contentserver.core
  :target-path "target/%s"
  :plugins [[lein-midje "3.1.3"]
            [lein-ring "0.8.13"]]
  :profiles {:uberjar {:dependencies [[javax.servlet/servlet-api "2.5"]
                                      [ring-mock "0.1.5"]]
                       :aot :all}}
  :omit-source true)


src/contentserver/core.clj
(ns contentserver.core (:gen-class)
  ^{:author "jabala"
    :doc "Bank Account API"
    :added "1.0"
    }
  (:require [clojure.tools.logging :as log]
            [contentserver.httpserver :as http]
            [contentserver.utils.config :as config]))

;; called when the program exits
(defn end-program
  "Shut application down gracefully"
  []
  ; stop server
  (http/stop-server ))

(defn -main
  "main entry of application"
  [& args]
  (log/info "===================================================")
  (log/info "BANK ACCOUNT API - VERSION 1.1.0")
  (log/info "===================================================")

  (config/load-configuration)                                 ; load configuration
  (http/initialize-http )                                   ; initialize http server

  ;; register runtime hook
  (.addShutdownHook (Runtime/getRuntime) (Thread. end-program))

  ;; start server
  (try
    (http/start-server)
    (catch Exception e
      (do
        (log/error "start server E: " (.getMessage e))
        ;; end program gracefully
        (end-program)))))

;(-main)




src/contentserver/utils/config.clj
(ns contentserver.utils.config
  ^{:author "joeabala"
    :doc "Load configurations"
    :added "1.0"
    }
  (:require [propertea.core :as propertea]
            [clojure.tools.logging :as log]))


;; configuration-file path
(def config (System/getProperty "bankaccount.config"))
;(def config "/Users/abala/Desktop/bankaccount/src/config.properties")

(defn get-configuration
  "Read from configuration file"
  [key]
  ((propertea/read-properties config)key))

(defn load-configuration
  "Load configuration file"
  []
  (log/infof "Loading configuration...")
  (def server-port (Integer/parseInt (get-configuration :server-port)))
  (def daily-max-deposit (Long/parseLong (get-configuration :daily-max-deposit)))            ;
  (def max-deposit-per-transact (Long/parseLong (get-configuration :max-deposit-per-transact)))
  (def max-deposit-freq (Integer/parseInt (get-configuration :max-deposit-freq)))
  (def daily-max-withdraw (Long/parseLong (get-configuration :daily-max-withdraw)))            ;
  (def max-withdraw-per-transact (Long/parseLong (get-configuration :max-withdraw-per-transact)))
  (def max-withdraw-freq (Integer/parseInt (get-configuration :max-withdraw-freq)))
  (log/infof "finished loading conf:  [%s %s %s %s %s %s %s]"
             server-port daily-max-deposit max-deposit-per-transact max-deposit-freq
             daily-max-withdraw max-withdraw-per-transact max-withdraw-freq))

;(load-configuration)

  
  

src/contentserver/model/datasource.clj
(ns contentserver.model.datasource
  ^{:author "joeabala"
    :doc    "Database configurations"
    :added  "1.0"
    }
  (:require
            [clojure.tools.logging :as log])
  (:use korma.core)
  (:use korma.db)
  (:use pg-hstore.core)
  (:import (java.sql SQLException)))

;==================================================================================================
;;                                  DATABASE SPECIFICATIONS
;==================================================================================================

(defdb db (postgres {:db         "bankaccount"
                     :user       "postgres"
                     :password   "root"
                     :host       "159.203.104.154"
                    ; :host       "localhost"
                     :port       "5432"
                     :delimiters ""}))

;==================================================================================================
;;                                    GENERAL INFORMATION
;==================================================================================================
(defn record-account
  "record new day transaction"
  [balance deposit withdrawal]
  (try
    (exec-raw db [(format "insert into tbl_account (balance, deposit_count, withdrawal_count)
                          values (%d,%d,%d) returning balance" balance deposit withdrawal)] :results)
    (catch SQLException e
      (log/error (.getMessage e)))
    (catch Exception e
      (log/error (.getMessage e)))))

(defn get-account-information
  "Get accoung information"
  []
  (try
    (exec-raw db ["select id, balance, deposit_count, withdrawal_count,
                  to_char(transaction_timestamp, 'dd-MM-yyyy')as transaction_timestamp from tbl_account order by transaction_timestamp desc limit 1"] :results)
    (catch SQLException e
      (log/error (.getMessage e)))
    (catch Exception e
      (log/error (.getMessage e)))))

(get-account-information)

(defn update-deposit-information
  "update deposit information"
  [id balance deposit-count]
  (try
    (exec-raw db [(format "update tbl_account set balance=%d, deposit_count=%d where id=%d RETURNING balance;" balance deposit-count id)] :results)
    (catch SQLException e
      (log/error (.getMessage e)))
    (catch Exception e
      (log/error (.getMessage e)))))

(defn update-withdrawal-information
  "update withdrawal information"
  [id balance withdraw-count]
  (try
    (exec-raw db [(format "update tbl_account set balance=%d, withdrawal_count=%d where id=%d returning balance" balance withdraw-count id)] :results)
    (catch SQLException e
      (log/error (.getMessage e)))
    (catch Exception e
      (log/error (.getMessage e)))))


src/contentserver/httpserver.clj
(ns contentserver.httpserver
  ^{:author "joeabala"
    :doc "HTTP Server"
    :added "1.0"
    }
  (:require [clojure.tools.logging :as log]
            [contentserver.utils.config :as config]
            [contentserver.model.datasource :as db]
            [clojure.data.json :as json]
            [ring.middleware.defaults :refer [wrap-defaults site-defaults]])
  (:use org.httpkit.server
        [compojure.route :only [not-found]]
        [compojure.handler :only [site]] ; form, query params decode; cookie; session, etc
        [compojure.core :only [defroutes GET POST DELETE ANY context]]))

(defn initialize-http
  "Define server once"
  []
  (defonce server (atom nil))
  (log/info "initializing http server ..."))

(def track-deposit-amount (atom 0)) ;; track state of deposit amount, to curb maximum daily deposit
(def track-withdraw-amount (atom 0)) ;; track state of withdrawal amount

(defn home
  "Welcome page"
  [req]
  {:status 200
   :headers {"Content-Type" "application/json; charset=utf-8"}
   :body (json/write-str {:status-code 0 :status-msg "Welcome to Bank Account API"})})

(defn process-balance-request
  [req]
  "Process balance request"
  (let [{balance :balance} (first (db/get-account-information)) ;; get balance from datastore
        balance (if (nil? balance) 0 balance)] ;; if there is no information returned, set balance as 0
    {:status 200
     :headers {"Content-Type" "application/json; charset=utf-8"}
     :body (json/write-str {:status-code 0 :status-msg "Your balance request is successful" :balance balance})}))

(defn process-deposit-request
  [{:keys [params] :as args}]
  "Process deposit request"
  (let [{deposit :deposit} params ;; get deposit amount from request
        today-date (.format (java.text.SimpleDateFormat. "dd-MM-yyyy") (new java.util.Date)) ;; get today's date
        db-result (first (db/get-account-information)) ;; get account information from datastore
        deposit (try ;; validate amount i.e make sure its a valid integer and its more than $9
                    (if (> (Long/parseLong deposit) 9) (Long/parseLong deposit) false) ;; return deposit value if successful
                   (catch ClassCastException e
                     (log/error (.getMessage e))
                     false ) ;; return false on exception
                    (catch NumberFormatException e
                      (log/error (.getMessage e))
                      false ) ;; return false on exception
                    (catch Exception e
                      (log/error (.getMessage e))
                      false ))] ;; return false on exception

    (if (not= false deposit) ;; check if deposit value is valid, if its valid proceed
      (if (< deposit config/max-deposit-per-transact) ;; check if deposit does not exceed daily max, if it does not proceed
        (if (seq db-result) ;; check if there is any data returned from datasource, if yes proceed
          (if  (< (+ deposit @track-deposit-amount) config/daily-max-deposit) ;; check if daily max deposit has been achieved, if not proceed
            (if (= (clojure.string/trim today-date) (clojure.string/trim (:transaction_timestamp db-result))) ;; check if this is first user transaction for the day, if not proceed
              (if (> config/max-deposit-freq (:deposit_count db-result)) ;; check if daily deposit limit has been reached, if not proceed
                (do
                  ;; all steps achieved, track deposit amount and update account
                  (log/info (:deposit_count db-result) " transaction for the day ")
                  (let [increment-deposit-amount-tracker (+ @track-deposit-amount deposit) ;; track deposit amount
                        deposit-amount-tracker (reset! track-deposit-amount increment-deposit-amount-tracker) ;; track deposit amount
                        _ (log/info "deposit-amount-tracker|depo-count " deposit-amount-tracker (:deposit_count db-result)) ;; log deposit amount and deposit count
                        {id :id} db-result ;; get record id
                        {balance :balance} db-result ;; get balance
                        {deposit-count :deposit_count} db-result ;; get deposit count
                        increment-daily-count (+ 1 deposit-count) ;; increment deposit by 1, to track deposit limit
                        _ (log/infof "{%s %s %s}" id balance deposit) ;; log id, balance and deposit
                        db-result (db/update-deposit-information id (+ balance deposit) increment-daily-count) ;; update account
                        {new-balance :balance} (first db-result);; get new balance
                        _ (log/info "new bal " new-balance)]
                    {:status 422
                     :headers {"Content-Type" "application/json; charset=utf-8"}
                     :body (json/write-str {:status-code 0 :status-msg "Your deposit was successful" :balance new-balance})}))
                (do
                  ;; daily deposit limit reached, return 422
                  {:status 422
                   :headers {"Content-Type" "application/json; charset=utf-8"}
                   :body (json/write-str {:status-code -2 :status-msg "Your daily deposit limit has reached" })}))
              (do
                ;; this is the first transaction for the day, record to database
                (log/info "new transaction for the day " deposit)
                (let [increment-deposit-amount-tracker (+ @track-deposit-amount deposit) ;; track deposit amount
                      _ (reset! track-deposit-amount increment-deposit-amount-tracker) ;; track deposit amount
                      db-result (db/record-account (+ deposit (:balance db-result)) 1 0) ;; load to database
                      {new-balance :balance} (first db-result);; get new balance
                      _ (log/info "new bal " new-balance)]
                  {:status 200
                   :headers {"Content-Type" "application/json; charset=utf-8"}
                   :body (json/write-str {:status-code 0 :status-msg "Your deposit was successful" :balance new-balance})})))
            (do
              ;; daily max deposit has been achieved, return 422
              {:status 422
               :headers {"Content-Type" "application/json; charset=utf-8"}
               :body (json/write-str {:status-code -2 :status-msg "You have reached your daily max deposit of $150,000"})})
            )
          (do
            ;; we dont have any data on database i.e account is inactive, record new data
            (let [increment-deposit-amount-tracker (+ @track-deposit-amount deposit) ;; calculate daily deposit amount
                  _ (reset! track-deposit-amount increment-deposit-amount-tracker) ;; track deposit amount
                  db-result (db/record-account deposit 1 0) ;; record data to datasource
                  {new-balance :balance} (first db-result)] ;; get new balance
              {:status 422
               :headers {"Content-Type" "application/json; charset=utf-8"}
               :body (json/write-str {:status-code -2 :status-msg "Your deposit was successful" :balance new-balance})})))
        (do
          ;; deposit is more than daily max deposit, return 422
          {:status 422
           :headers {"Content-Type" "application/json; charset=utf-8"}
           :body (json/write-str {:status-code -2 :status-msg "You cannot deposit more than $40,000"})}))
      (do
        ;; invalid deposit, return 422
        {:status 422
         :headers {"Content-Type" "application/json; charset=utf-8"}
         :body (json/write-str {:status-code -2 :status-msg "Invalid Amount"})}))))


(defn process-withdrawal-request
  [{:keys [params] :as args}]
  (log/info "my request " args)
  (let [{withdraw :withdraw} params  ;; get withdraw amount from request
        today-date (.format (java.text.SimpleDateFormat. "dd-MM-yyyy") (new java.util.Date)) ;; get today's date
        db-result (first (db/get-account-information)) ;; get account information
        withdraw (try ;; validate amount i.e make sure its a valid integer and its more than $9
                  (if (> (Long/parseLong withdraw) 9) (Long/parseLong withdraw) false);; return deposit value if successful
                  (catch ClassCastException e
                    (log/error "::"(.getMessage e))
                    false ) ;; return false on exception
                  (catch NumberFormatException e
                    (log/error (.getMessage e))
                    false ) ;; return false on exception
                  (catch Exception e
                    (log/error (.getMessage e))
                    false ))] ;; return false on exception
    (if (not= false withdraw) ;; check if withdraw value is valid, if its valid proceed
      (if (< withdraw config/max-withdraw-per-transact);; check if withdrawal does not exceed daily max, if it does not proceed
        (if (seq db-result) ;; check if there is any account information in datastore
          (if  (< (+ withdraw @track-withdraw-amount) config/daily-max-withdraw) ;;check if daily max withdrawal has been achieved, if not proceed
            (if (= (clojure.string/trim today-date) (clojure.string/trim (:transaction_timestamp db-result))) ;; check if this is first user transaction for the day, if not proceed
              (if (> config/max-withdraw-freq (:withdrawal_count db-result))
                (let [increment-withdraw-amount-tracker (+ @track-withdraw-amount withdraw);; track withdrawal amount
                      withdraw-amount-tracker (reset! track-withdraw-amount increment-withdraw-amount-tracker) ;; track withdrawal amount
                      _ (log/info "withdraw-amount-tracker3|depo-count " withdraw-amount-tracker (:withdrawal_count db-result))
                      {id :id} db-result ;; get record id
                      {balance :balance} db-result ;; get balance
                      {withdraw-count :withdrawal_count} db-result ;; get withdrawal count
                      increment-daily-count (+ 1 withdraw-count) ;; increment withdrawal count
                      db-result (db/update-withdrawal-information id (- balance withdraw) increment-daily-count) ;; update account
                      {new-balance :balance} (first db-result)] ;; get new balance
                  {:status 200
                   :headers {"Content-Type" "application/json; charset=utf-8"}
                   :body (json/write-str {:status-code 0 :status-msg (str "You have withdrawn $" withdraw) :balance new-balance})})
                {:status 200
                 :headers {"Content-Type" "application/json; charset=utf-8"}
                 :body (json/write-str {:status-code 0 :status-msg "Your daily withdrawal limit has reached"})}
                )
              (do
                ;; first user withdrawal for the day, record account info
                (let [increment-withdraw-amount-tracker (+ @track-withdraw-amount withdraw) ;; track withdrawal amount
                      withdraw-amount-tracker (reset! track-withdraw-amount increment-withdraw-amount-tracker);; track withdrawal amount
                      _ (log/info "withdraw-amount-tracker2 " withdraw-amount-tracker)
                      db-result (db/record-account (- (:balance db-result) withdraw) 0 1) ;; record account information
                      {new-balance :balance} (first db-result)] ;; get new balance
                  {:status 200
                   :headers {"Content-Type" "application/json; charset=utf-8"}
                   :body (json/write-str {:status-code 0 :status-msg (str "You have withdrawn $" withdraw) :balance new-balance})})))
            (do
              ;; you have exceeded your daily withdrawal
              {:status 422
               :headers {"Content-Type" "application/json; charset=utf-8"}
               :body (json/write-str {:status-code -2 :status-msg "You have reached your daily max withdrawal of $50,000"})}))
          (do
            ;; there is no account information, please make deposit
            {:status 422
             :headers {"Content-Type" "application/json; charset=utf-8"}
             :body (json/write-str {:status-code -2 :status-msg "Please activate your account"})}))
        (do
          ;; withdrawal is more than daily max withdrawal, return 422
          {:status 422
           :headers {"Content-Type" "application/json; charset=utf-8"}
           :body (json/write-str {:status-code -2 :status-msg "You cannot withdraw more than $20,000"})}))
      (do
        ;; invalid withdrawal, return 422
        {:status 422
         :headers {"Content-Type" "application/json; charset=utf-8"}
         :body (json/write-str {:status-code -2 :status-msg "Invalid Amount"})}))))


;; routes
(defroutes all-routes
           (GET "/" [] home)
           (GET "/balance" [] process-balance-request)
           (POST "/deposit" [] process-deposit-request)
           (POST "/withdraw" [] process-withdrawal-request)
           (not-found {:status 404
                       :headers {"Content-Type" "application/json; charset=utf-8"}
                       :body (json/write-str {:status-code -1 :status-msg "Invalid Request"})})) ;; return 404


(def app
  (wrap-defaults all-routes (assoc-in site-defaults [:security :anti-forgery] false)))

(defn start-server
  "Start server"
  []
  (try
    (do
      (reset! server (run-server (site #'all-routes) {:port config/server-port}))
      (log/info "server started at port " config/server-port))
    (catch Exception e
      (do
        (throw (Exception. (str "Error: port " config/server-port " is already in use. " e)))))))

(defn stop-server
  "Gracefuly shut down server"
  []
  (when-not (nil? @server)
    (@server :timeout 100) ;; wait 100ms for existing requests to be finished
    (reset! server nil)
    (log/info "server shut down")))


